session/session.go

```diff
@@ -185,6 +185,76 @@ type session struct {
 	statsCollector *handle.SessionStatsCollector
 	// ddlOwnerChecker is used in `select tidb_is_ddl_owner()` statement;
 	ddlOwnerChecker owner.DDLOwnerChecker
+	// lockedTables use to record the table locks hold by the session.
+	lockedTables struct {
+		sync.RWMutex
+		holdLocks map[int64]model.TableLockTpInfo
+	}
+}
+
+// AddTableLock adds table lock to the session lock map.
+func (s *session) AddTableLock(locks []model.TableLockTpInfo) {
+	s.lockedTables.Lock()
+	for _, l := range locks {
+		s.lockedTables.holdLocks[l.TableID] = l
+	}
+	s.lockedTables.Unlock()
+}
+
+// ReleaseTableLocks releases table lock in the session lock map.
+func (s *session) ReleaseTableLocks(locks []model.TableLockTpInfo) {
+	s.lockedTables.Lock()
+	for _, l := range locks {
+		delete(s.lockedTables.holdLocks, l.TableID)
+	}
+	s.lockedTables.Unlock()
+}
+
+// ReleaseTableLockByTableIDs releases table lock in the session lock map by table ID.
+func (s *session) ReleaseTableLockByTableIDs(tableIDs []int64) {
+	s.lockedTables.Lock()
+	for _, tblID := range tableIDs {
+		delete(s.lockedTables.holdLocks, tblID)
+	}
+	s.lockedTables.Unlock()
+}
+
+// CheckTableLocked checks the table lock.
+func (s *session) CheckTableLocked(tblID int64) (bool, model.TableLockType) {
+	s.lockedTables.RLock()
+	defer s.lockedTables.RUnlock()
+	lt, ok := s.lockedTables.holdLocks[tblID]
+	if !ok {
+		return false, model.TableLockNone
+	}
+	return true, lt.Tp
+}
+
+// GetAllTableLocks gets all table locks table id and db id hold by the session.
+func (s *session) GetAllTableLocks() []model.TableLockTpInfo {
+	s.lockedTables.RLock()
+	lockTpInfo := make([]model.TableLockTpInfo, 0, len(s.lockedTables.holdLocks))
+	for _, tl := range s.lockedTables.holdLocks {
+		lockTpInfo = append(lockTpInfo, tl)
+	}
+	s.lockedTables.RUnlock()
+	return lockTpInfo
+}
+
+// HasLockedTables uses to check whether this session locked any tables.
+// If so, the session can only visit the table which locked by self.
+func (s *session) HasLockedTables() bool {
+	s.lockedTables.RLock()
```

Comment:
Could we just use an atomic value to indicate the session has locked tables?

Reply:
Maintenance another `atomic status value` to keep sync with the `lockedTables`? I don't think it's a good idea.

Reply:
I will use double check lock in here in next pr.